pkg/typesystem/typesystem.go

```diff
@@ -781,7 +781,9 @@ func (t *TypeSystem) validateRelationTypeRestrictions() error {
 			if assignable && len(relatedTypes) == 1 {
 				relatedObjectType := relatedTypes[0].GetType()
 				relatedRelation := relatedTypes[0].GetRelation()
-				if objectType == relatedObjectType && name == relatedRelation {
+
+				if objectType == relatedObjectType && name == relatedRelation &&
+					reflect.TypeOf(relation.GetRewrite().Userset) == reflect.TypeOf(&openfgapb.Userset_This{}) {
```

Comment:
We do not, because the relation is loaded dynamically at runtime. We lookup the relation and based on its rewrite's type (which is determined at runtime) we do some validation. So this behavior is at runtime, not statically at compile time.

Reply:
As discussed for now, we can proceed like this for now.